### DISCUSSION TITLE
Count number of unique colors on the image to see if it's cartoon-like

### DISCUSSION THREAD

**Discussion by dooman87** [ImageMagick/ImageMagick#4315] [Digital Image Processing]
Hi,

I'm trying to write a function that would classify if the image a cartoon or not. I found the [algorithm](https://legacy.imagemagick.org/Usage/compare/#type_reallife) in the documentation:

> Jim Van Zandt offers this solution...
> write out the color of every pixel
> sort by color
> write out the pixel count for every color
> sort by pixel count
> Work your way through the list until you have accounted for half the pixels in the image.
> If #pixels >>> #colors then it's cartoon like.

I tried to use `histogram:info`, but it was taking too much time. So I tried to write my own implementation using MagicWand and `PixelRegionIterator`. The code of the function is here https://github.com/Pixboost/transformimgs/blob/feature/find-photo/img/processor/imagemagick.go#L294. 

The code above goes through the each pixel of the image and increase counter. I played with the different versions of implementation and the fastest was to use go routines (threads) with lock-free map.

Currently it takes 2.1 secs on my laptop (Dell XPS) to go through 3200x2200 image and the 90% of the time is iterating through the pixels.

Is there any ways to solve the above problem or improve the algorithm so it takes less time?

**Comment by GeeMack** [comment]
Using ImageMagick v6 on the command line you can get the ratio of unique colors to total pixels with a command like this...

    convert input.png ( +clone -unique-colors ) -set option:result "%[fx:v.w/(u.w*u.h)]" +delete -format "%[result]\n" info:

That reads in the image, and inside the parentheses it creates an image 1 pixel high that contains only one of each unique color. Then it uses an FX expression to calculate the number of unique colors, "`v.w`", divided by the total input image pixels, "`(u.w*u.h)`", and outputs the result as text. You can work the calculations in those FX expressions many different ways to produce the exact results you need.

That should work with any input image on a Windows command line. For a *nix OS you'll need to escape the parentheses "(...)" with backslashes "\\(...\\)", and maybe escape  any other special characters. For IMv7 you'll use "magick" instead of "convert".

**Reply by dooman87** [reply]
Thanks a lot for that. It's actually much faster and almost covering all test cases!

It does fail on some images with background, e.g. ecommerce product image with monotone background. I excluded the first (most used) color if it fills more than 10% of the image in my code. I would also exclude all the background pixels from ratio calculation. 

Do you think by any chance there is some magic I can add to your example to do similar?

**Comment by snibgo** [comment]
For images with a monotone background, you would want to (a) subtract one from the number of unique colours and (b) subtract the number of background pixels from the width*height.

If you turn the background black and all other pixels white, %[fx:mean\*w\*h] is the number of non-background pixels.

**Comment by fmw42** [comment]
Snibgo, I edited your post to put \ before *, so that they would show properly. I hope you do not mind.

**Comment by snibgo** [comment]
Aah, stupid me, thanks, fmw42. It's only been a few days since I forgot to escape backslashes. Note to self: always check the post after posting.

**Comment by dooman87** [comment]
Thanks a lot everyone, it's a very good idea with subtracting background.

Turned out using ImageGetHistogram is about 4-5 times faster than using `histogram:info:` from CLI and the performance is at sufficient level, I think. 

Just, in case anyone interested here is the code I got so far. The only issue I noticed so far is screenshots, but I think we can go with those being classified as cartoons:

```go
// IsIllustration return true if image is cartoon like, including
// icons, illustrations. However, banners are not included.
//
// The initial idea is from here: https://legacy.imagemagick.org/Usage/compare/#type_reallife
func (p *ImageMagick) IsIllustration(src *img.Image) (bool, error) {
	start := time.Now()

	mw := imagick.NewMagickWand()
	err := mw.ReadImageBlob(src.Data)
	if err != nil {
		return false, err
	}
	fmt.Printf("[%s] Read image: %d\n", src.Id, time.Since(start).Milliseconds())

	_, colors := mw.GetImageHistogram()
	fmt.Printf("Get histogram: %d\n", time.Since(start).Milliseconds())

	sort.Slice(colors, func(i, j int) bool {
		return colors[i].GetColorCount() > colors[j].GetColorCount()
	})

	imageWidth := int(mw.GetImageWidth())
	imageHeight := int(mw.GetImageHeight())
	totalPixelsCount := imageHeight * imageWidth

	var colorIdx int
	var c *imagick.PixelWand
	background := 0
	pixels := 0
	tenPercent := int(float32(totalPixelsCount) * 0.1)
	fiftyPercent := int(float32(totalPixelsCount) * 0.5)

	for colorIdx, c = range colors {
		count := int(c.GetColorCount())
		if colorIdx == 0 && count >= tenPercent {
			background = count
			fiftyPercent = int((float32(totalPixelsCount) - float32(background)) * 0.5)
			continue
		}

		if pixels > fiftyPercent {
			break
		}

		pixels += count
	}
	fmt.Printf("Colors Iteration: %d\n", time.Since(start).Milliseconds())

	fmt.Printf("[%d] of [%d] with pixels = [%d]\n", colorIdx, len(colors), fiftyPercent)

	return colorIdx*500 < fiftyPercent, nil
}

```

**Comment by fmw42** [comment]
Cartoons or line drawings that are mostly large regions of solid colors would have a low standard deviation. Natural images with lots of changing texture would have a high standard deviation. Have you tried simply thresholding on the standard deviation to determine if cartoons or not?

**Reply by dooman87** [reply]
Hi Fred,

Thanks a lot for that! I didn't know about standard deviation, so that was a very useful learning! I think I could've used it in case of line drawning, but I would also like to separate illustrations, from banners. I feel like if maybe I could remove background then standard deviation might work better. Below are some examples:

Illustration (should return `true`), standard deviation: 11697.1 (0.178487)
![people](https://user-images.githubusercontent.com/124110/136369414-0490ce77-dc50-41ec-b2c9-785b36703e46.png)

Icon (should return `true`), standard deviation: 111.32 (0.436547)
![image-api](https://user-images.githubusercontent.com/124110/136369525-8306f279-d043-4029-9a88-48bcb65612ae.png)

Product image (should return `false`), standard deviation: 126.647 (0.496654)
![markers](https://user-images.githubusercontent.com/124110/136369543-f2ff4663-5d84-4e01-9aad-d6c13948919c.png)

Banner (should return `false), standard deviation: 71.8775 (0.281872)
![l-carnitine-shaker-banner](https://user-images.githubusercontent.com/124110/136369943-23705d3f-bb57-4a51-a78e-0f7f8cdda35e.png)

**Comment by Jmuccigr** [comment]
Spitballing here, but you might try a trim with a little fuzz before doing the analysis to get rid of lots of monocolor background that may skew the results.

**Comment by fmw42** [comment]
You have graphics, which can have lots of changes of colors. In that case standard deviation will increase and may not be a good measure.

Perhaps a better measure would be to do  edge detection and then get the density of edges via the image average.


convert illustration.png -canny 0x1+5%+10% +write x1.png -format "%[fx:100*mean]\n" info:
0.6215

convert icon.png -canny 0x1+5%+10% +write x2.png -format "%[fx:100*mean]\n" info:
1.38478

convert product.png -canny 0x1+5%+10% +write x3.png -format "%[fx:100*mean]\n" info:
7.10375


Included is a +write x.png so that you can view the edge images.

Does that work better?

Fred

**Reply by dooman87** [reply]
O! this one looks very promising so far. I had to decrease blur a bit, but it seems to work much better than the previous one. I'm going to run it on the bigger dataset and see if there are any problems. Thanks a lot for your help!

**Comment by dooman87** [comment]
Just a quick update on this one. I was pursuing the original path and stumbled upon the problem with memory consumption of `MagickGetImageHistogram()`. For the image below it eats 1Gb of memory which is quite high and not going to work for my use case. 

![3](https://user-images.githubusercontent.com/124110/138186282-2fbcba00-5f55-46a3-b2f1-d833387cde22.png)

@fmw42 The edge detection is working very well when running with less blur, but sometime takes quite a long time (3-4 secs), so would need to evaluate if that's gonna fly for our use case.

@Jmuccigr Going to try do some fuzzing and count number of colors. I have a feeling it might work, but will need to run tests on our dataset.

**Reply by jyrkialakuijala** [reply]
When I count colors in WebP encoding to decide about palette encoding, I used a fixed size (1024 entries) hash table. When I observe more than 256 colors, I stop the computation, so the hash table never gets more than 25 % full. 

A fixed sized hash table is relatively easy to implement. Perhaps a similar approach would work here.
